src/core/PasswordGenerator.cpp

```diff
@@ -52,47 +53,99 @@ QString PasswordGenerator::generatePassword() const
 {
     Q_ASSERT(isValid());
 
-    const QVector<PasswordGroup> groups = passwordGroups();
+    const QVector<PasswordGroup> groups_original = passwordGroups();
+    QVector<PasswordGroup> groups = groups_original;
 
-    QVector<QChar> passwordChars;
+    QVector<QChar> passwordChars_original;
     for (const PasswordGroup& group : groups) {
         for (QChar ch : group) {
-            passwordChars.append(ch);
+            passwordChars_original.append(ch);
         }
     }
+    QVector<QChar> passwordChars = passwordChars_original;
 
     QString password;
+    //qDebug().noquote().nospace() << "*** Begin generating password ***";
 
     if (m_flags & CharFromEveryGroup) {
-        for (int i = 0; i < groups.size(); i++) {
-            int pos = randomGen()->randomUInt(groups[i].size());
-
-            password.append(groups[i][pos]);
-        }
 
-        for (int i = groups.size(); i < m_length; i++) {
-            int pos = randomGen()->randomUInt(passwordChars.size());
-
-            password.append(passwordChars[pos]);
+        /*
+         * Continuously add one random character from each group until pass-
+         * word has reached the desired length
+         *
+         * NOTE:
+         * Same as below, but more readable, just for reference
+         */
+        /*while (password.length() < m_length) {
+            for (int i = 0; i < groups.size(); i++) {
+                int pos = randomGen()->randomUInt(groups[i].size());
+                password.append(groups[i][pos]);
+
+                if (m_flags & MinimizeRecurrence) {
+                    groups[i].remove(pos);
+                    if (groups[i].isEmpty()) {
+                        groups[i] = groups_original[i];
+                    }
+                }
+
+                if (password.length() >= m_length) {
+                    break;
+                }
+            }
+        }*/
+
+        /*
+         * Continuously add one random character from each group until pass-
+         * word has reached the desired length
+         *
+         * NOTE:
+         * Optimized variant of above algorithm
+         */
+        for (int i = 0; i < m_length; i++) {
+            int group = i % groups.size();
+            int pos = randomGen()->randomUInt(groups[group].size());
```

Comment:
I really don't like either approach, because they both decrease entropy. In the original approach, the alphabet of the first `groups.size()` characters is reduced to their group, in your approach the alphabet of each character in the password is reduced to the alphabet of `i % groups.size()` (which is worse).

The correct algorithm would be this (pseudo code):

```
forceGroupAt = Map()
for i in 0 -> groups.size() do:
    forceGroupAt.put(randInt(0, passwordLength) not in forceGroupAt, i)

for i in 0 -> passwordLength do:
    if pickCharFromEveryGroup and i in forceGroupAt then:
        password.append(randCharFromGroup(forceGroupAt.get(i))
    else:
        password.append(randCharFromSelectedGroups())
```

That way you guarantee at least one random character from every group at a random position.